src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -1242,92 +1252,105 @@ internal static bool DeepEquals(JsonElement left, JsonElement right)
+                        return false;
+                    }
+
+                    ArrayEnumerator arrayEnumerator2 = element2.EnumerateArray();
+                    foreach (JsonElement e1 in element1.EnumerateArray())
                     {
-                        if (!rightArrayEnumerator.MoveNext() || !DeepEquals(leftElement, rightArrayEnumerator.Current))
+                        if (!arrayEnumerator2.MoveNext() || !DeepEquals(e1, arrayEnumerator2.Current))
                         {
                             return false;
                         }
                     }
 
-                    return !rightArrayEnumerator.MoveNext();
+                    Debug.Assert(!arrayEnumerator2.MoveNext());
+                    return true;
 
                 default:
                     Debug.Assert(kind is JsonValueKind.Object);
-                    ObjectEnumerator leftObjectEnumerator = left.EnumerateObject();
-                    ObjectEnumerator rightObjectEnumerator = right.EnumerateObject();
+
+                    int count = element1.GetObjectSize();
+                    if (count != element2.GetObjectSize())
+                    {
+                        return false;
+                    }
+
+                    ObjectEnumerator objectEnumerator1 = element1.EnumerateObject();
+                    ObjectEnumerator objectEnumerator2 = element2.EnumerateObject();
 
                     // Two JSON objects are considered equal if they define the same set of properties.
                     // Start optimistically with sequential comparison, but fall back to unordered
                     // comparison as soon as a mismatch is encountered.
 
-                    while (leftObjectEnumerator.MoveNext())
+                    while (objectEnumerator1.MoveNext())
                     {
-                        if (!rightObjectEnumerator.MoveNext())
+                        if (!objectEnumerator2.MoveNext())
                         {
                             return false;
                         }
 
-                        JsonProperty leftProp = leftObjectEnumerator.Current;
-                        JsonProperty rightProp = rightObjectEnumerator.Current;
+                        JsonProperty prop1 = objectEnumerator1.Current;
+                        JsonProperty prop2 = objectEnumerator2.Current;
 
-                        if (!NameEquals(leftProp, rightProp))
+                        if (!NameEquals(prop1, prop2))
                         {
                             // We have our first mismatch, fall back to unordered comparison.
-                            return UnorderedObjectDeepEquals(leftObjectEnumerator, rightObjectEnumerator);
+                            return UnorderedObjectDeepEquals(objectEnumerator1, objectEnumerator2, remainingProps: count);
                         }
 
-                        if (!DeepEquals(leftProp.Value, rightProp.Value))
+                        if (!DeepEquals(prop1.Value, prop2.Value))
```

Comment:
Do we need to be concerned about unbounded recursion, stack overflows, etc.?

Reply:
Good question, I would expect this to be handled by the `JsonDocument` parser, but will double check to be sure.

Reply:
The maximum depth of a `JsonDocument` is bounded by `JsonDocumentOptions.MaxDepth` which defaults to 64. Assuming the user manually sets `MaxDepth` to a large enough number, then `DeepEquals` can potentially stack overflow which is consistent with how `JsonSerializer` behaves.